src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,170 @@
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn load_milan_cert_chain() -> Result<(x509::X509, x509::X509)> {
+
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ark_ask.pem"))?;
+    
+    // ark, ask
+    return Ok((certs[0].clone(), certs[1].clone()))
+
+}
+
+fn verify_cert_chain(cert_chain: &CertTable) -> Result<x509::X509> {
+
+    let (ark, ask) = load_milan_cert_chain()?;
+
+    //let raw_vcek = cert_chain.entries.iter().find(|c| c.cert_type == SnpCertType::VCEK)
+    let raw_vcek = cert_chain
+        .entries
+        .iter()
+        .find(|c| c.guid() == "63da758d-e664-4564-adc5-f4b93be8accd")
+        .ok_or(anyhow!("VCEK not found."))?;
+    let vcek =
+        x509::X509::from_der(&raw_vcek.data()).map_err(|_| anyhow!("Failed to load VCEK"))?;
+
+    // ARK -> ASK
+    ask.verify(&(ark.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ASK Signature"))?;
+
+    // ASK -> VCEK
+    vcek.verify(&(ask.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid VCEK Signature"))?;
+
+    Ok(vcek)
+}
+
+fn calculate_expected_report_data(nonce: &String, tee_pubkey: &String) -> [u8; 64] {
+    let mut hasher = Sha384::new();
+
+    hasher.update(nonce.as_bytes());
+    hasher.update(tee_pubkey.as_bytes());
+
+    let mut hash = hasher.finalize().to_vec();
+    hash.extend([0u8; 16]);
+
+    hash.as_slice().try_into().unwrap()
+}
+
+fn parse_tee_evidence(report: &AttestationReport) -> Result<TeeEvidenceParsedClaim> {
+    let claims_map = json!({
+        // policy fields
+        "policy_abi_major": report.policy.abi_major(),
+        "policy_abi_minor": report.policy.api_minor(),
+        "policy_smt_allowed": report.policy.smt_allowed(),
+        "policy_migrate_ma": report.policy.migrate_ma_allowed(),
+        "policy_debug_allowed": report.policy.debug_allowed(),
+        "policy_single_socket": report.policy.single_socket_required(),
+
+        // versioning info
+        "current_tcb_bootloader": report.current_tcb.boot_loader,
+        "current_tcb_tee": report.current_tcb.tee,
+        "current_tcb_snp": report.current_tcb.snp,
+        "current_tcb_microcode": report.current_tcb.microcode,
+        "current_build": report.current_build,
+        "current_major": report.current_major,
+        "current_minor": report.current_minor,
+
+        // platform info
+        "platform_tsme_enabled": report.plat_info.tsme_enabled(),
+        "platform_smt_enabled": report.plat_info.smt_enabled(),
+
+        // measurement
+        "measurement": base64::encode(report.measurement),
+    });
+
+    Ok(claims_map as TeeEvidenceParsedClaim)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use openssl::nid::Nid;
+
+    #[test]
+    fn check_milan_certificates() {
+        let (ark, ask) = load_milan_cert_chain().unwrap();
+        assert_eq!(get_common_name(ark).unwrap(), "SEV-Milan");
+        assert_eq!(get_common_name(ask).unwrap(), "ARK-Milan");
```

Comment:
So I split the pem file into 1.pem (first) and 2.pem (second) certs. Then I looked at them:

```
$ openssl x509 -in 2.pem -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 65536 (0x10000)
        Signature Algorithm: rsassaPss
         Hash Algorithm: sha384
         Mask Algorithm: mgf1 with sha384
          Salt Length: 0x30
         Trailer Field: 0x01
        Issuer: OU = Engineering, C = US, L = Santa Clara, ST = CA, O = Advanced Micro Devices, CN = ARK-Milan
        Validity
            Not Before: Oct 22 17:23:05 2020 GMT
            Not After : Oct 22 17:23:05 2045 GMT
        Subject: OU = Engineering, C = US, L = Santa Clara, ST = CA, O = Advanced Micro Devices, CN = ARK-Milan
```

this (2.pem) looks like a self-signed cert (issuer = subject), and its subject is `ARK-Milan`.

as for the first one:

```
$ openssl x509 -in 1.pem -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 65537 (0x10001)
        Signature Algorithm: rsassaPss
         Hash Algorithm: sha384
         Mask Algorithm: mgf1 with sha384
          Salt Length: 0x30
         Trailer Field: 0x01
        Issuer: OU = Engineering, C = US, L = Santa Clara, ST = CA, O = Advanced Micro Devices, CN = ARK-Milan
        Validity
            Not Before: Oct 22 18:24:20 2020 GMT
            Not After : Oct 22 18:24:20 2045 GMT
        Subject: OU = Engineering, C = US, L = Santa Clara, ST = CA, O = Advanced Micro Devices, CN = SEV-Milan
```

This is signed by the ARK, so this looks to me like the ASK (even though the subject doesn't say ASK).

So are you sure about the order of the certs in the pem file?